include/zephyr/drivers/gpio.h

```diff
@@ -752,6 +754,92 @@ static inline int gpio_pin_configure_dt(const struct gpio_dt_spec *spec,
 				  spec->dt_flags | extra_flags);
 }
 
+/*
+ * @brief Get direction of all pins in a port.
+ *
+ * If @p inputs or @p outputs is NULL, then this function does not get the
+ * respective input or output direction information.
+ *
+ * @param port Pointer to the device structure for the driver instance.
+ * @param inputs Pointer to a variable where input directions will be stored.
+ * @param outputs Pointer to a variable where output directions will be stored.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO I/O error when accessing an external GPIO chip.
+ * @retval -EWOULDBLOCK if operation would block.
+ */
+__syscall int gpio_port_get_direction(const struct device *port, gpio_port_pins_t *inputs,
+					   gpio_port_pins_t *outputs);
+
+static inline int z_impl_gpio_port_get_direction(const struct device *port,
+						      gpio_port_pins_t *inputs,
+						      gpio_port_pins_t *outputs)
+{
+	const struct gpio_driver_api *api = (const struct gpio_driver_api *)port->api;
+
+	if (api->port_get_direction_raw == NULL) {
+		return -ENOTSUP;
```

Comment:
I think the correct return value here would be `-ENOSYS` (see https://docs.zephyrproject.org/latest/develop/api/design_guidelines.html#return-codes).